### PR TITLE
Dockerfile

### PR DESCRIPTION
**Pull request by Bundaberg-Joey**
## Description
This PR adds a Dockerfile which allows for easy installation of all python dependencies, the pyscreener library, ADFR suite tools and choice of vina based docking software to a contained docker image. This allows for reproducible and straightforward installation / testing / usage of the pyscreener code. All pytest checks pass as well as `smoke-test` for all vina versions.

## Example / Current workflow
Creating a docker image containing the pyscreener library, its dependencies, ADFR suite tools and the vina docking software can be accomplished with `docker build -t pyscreener:vina --target vina .` however any of the vina base docking softwares can be specified instead.

## Bugfix / Desired workflow
n/a

## Questions
Installing `qvina` / `psovina` requires building from source which raises some pragma issues during compilation. If any light can be shed on solving these issues / clarifying if these issues then that would be invaluable feedback.

## Relevant issues
n/a

## checklist
No new tests added but all docker builds pass the associated python and docking smoke tests

## Status
Not 100% ready as seeking feedback on compilation of `qvina` and `psovina` docking tools from souce

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -5,9 +5,6 @@ build-backend = "setuptools.build_meta"
 [tool.setuptools_scm]
 write_to = "pyscreener/_version.py"
 
-[tool.pytest.ini_options]
-addopts = "--cov pyscreener"
-
```
Comment (davidegraff) [inline pyproject.toml:10]: why is this removed?

**Review by davidegraff — COMMENTED** [review]
What great timing on this- I was just thinking the other day about how I should finally go about making the Dockerfile. It all looks great to me. A few questions:
- is it possible to add a base image for DOCK6? I know the source download requires a license, so we can't install and build automatically, but there are a few additional utilities that we could include a base DOCK6 image
- how do you feel about using micromamba instead of miniconda? In my experience, it's usually interchangeable with conda and results in smaller images with faster builds

File: pyproject.toml (hunk 1 above)
Comment (Bundaberg-Joey) [reply pyproject.toml]: My apologies, I'd removed that during testing as I was having issues with pytest. Never meant for it to be committed which shows the value of the PR system I guess!

I've updated the readme to specify pip installation of `pytest-cov` instead of just `pytest` as the standard version doesn't conduct coverage checks.

**Comment by Bundaberg-Joey** [comment]
> is it possible to add a base image for DOCK6? I know the source download requires a license, so we can't install and build automatically, but there are a few additional utilities that we could include a base DOCK6 image

Sure, from the README I'm assuming the desired software would be [sphgen_cpp](http://dock.compbio.ucsf.edu/Contributed_Code/sphgen_cpp.htm) and [chimera](https://www.cgl.ucsf.edu/chimera/download.html)? 

Seeing as DOCK6 doesn't rely on the ADFR suite, I'd imagine re-structuring the final dockerfile as below:

```bash
# base  (python dependencies only)
FROM continuumio/miniconda3 AS base

# vina base (installs ADFR)
FROM base AS vina-base

# dock6 base (installs relevant dock6 accessories)
FROM base AS dock6-base

# final vina flavour images (specific vina docking softwares_
FROM vina-base AS psovina
... etc
``` 
If you could suggest a better name than `vina-base` however I'd be all for it to avoid ambiguities in the naming (i.e. vina vs vina-base)
 
> how do you feel about using micromamba instead of miniconda? In my experience, it's usually interchangeable with conda and results in smaller images with faster builds

I'd never say no to a smaller or faster build let alone both simultaneously! I'll start working on creating the docker file as above from the miniconda image then move over to the micromamba version. From what you say should hopefully be a simple change at the very top!

**Comment by davidegraff** [comment]
yeah i think that new organization makes sense re: DOCK6. Of course, you can't add `sphgen_cpp` to the `$DOCK6/bin` because it doesn't exist, but at least it'll be installed in the image

**Comment by Bundaberg-Joey** [comment]
@davidegraff  have successfully managed to swap out the conda base image for the micromamba base image.
So far as I can see there's nothing left outstanding with this PR.

File: README.md [hunk 2]
```diff
@@ -23,16 +23,25 @@ This repository contains the source of pyscreener, both a library and software f
...
 * `docker build -t pyscreener:psovina --target psovina .` : Creates an image from `pyscreener:base` with `psovina` installed 
 
-If docker is not installed already for your system then it can be installed from [the official docker website](https://docs.docker.com/get-docker/).
+As `DOCK6` software requires a license, it is not possible to include its installation within the associated docker image.
+A compiled form of `sphgen_cpp` and the binary required for installation of `chimera` are both available within the `dock6_utils` directory of the associated dock6 imge:
+* `docker build -t pyscreener:dock6 --target base-dock6 .` : Creates an image from `pyscreener:base` containing utility software needed for `DOCK6` to run once installed 
 
-Note : The above commands assume you are running them within the directory containing both `Dockerfile` and `environment.yml`
+Notes : 
+1. Within the docker container, the environment `base` will e activated by default which contains all the required python dependencies so there is no need to manually activate an environment once inside the container
```
Comment (davidegraff) [inline README.md:43]:
```suggestion
1. Within the docker container, the environment `base` will be activated by default. This contains all the required python dependencies so there is no need to manually activate an environment once inside the container
```

File: README.md [hunk 3]
```diff
@@ -23,16 +23,25 @@ This repository contains the source of pyscreener, both a library and software f
...
 * `docker build -t pyscreener:smina --target smina .` : Creates an image from `pyscreener:base` with `smina` installed 
 * `docker build -t pyscreener:psovina --target psovina .` : Creates an image from `pyscreener:base` with `psovina` installed 
 
-If docker is not installed already for your system then it can be installed from [the official docker website](https://docs.docker.com/get-docker/).
+As `DOCK6` software requires a license, it is not possible to include its installation within the associated docker image.
+A compiled form of `sphgen_cpp` and the binary required for installation of `chimera` are both available within the `dock6_utils` directory of the associated dock6 imge:
+* `docker build -t pyscreener:dock6 --target base-dock6 .` : Creates an image from `pyscreener:base` containing utility software needed for `DOCK6` to run once installed 
 
-Note : The above commands assume you are running them within the directory containing both `Dockerfile` and `environment.yml`
+Notes :
```
Comment (davidegraff) [inline README.md:42]:
```suggestion
**Notes :** 
```

File: README.md [hunk 4]
```diff
@@ -23,16 +23,25 @@ This repository contains the source of pyscreener, both a library and software f
...
+* `docker build -t pyscreener:base --target base .` : Creates a docker image containing all python dependencies and pyscreener library but no docking software 
 * `docker build -t pyscreener:vina --target vina .` : Creates an image from `pyscreener:base` with `vina` installed 
 * `docker build -t pyscreener:qvina --target qvina .` : Creates an image from `pyscreener:base` with `qvina` installed 
 * `docker build -t pyscreener:smina --target smina .` : Creates an image from `pyscreener:base` with `smina` installed 
 * `docker build -t pyscreener:psovina --target psovina .` : Creates an image from `pyscreener:base` with `psovina` installed 
 
-If docker is not installed already for your system then it can be installed from [the official docker website](https://docs.docker.com/get-docker/).
+As `DOCK6` software requires a license, it is not possible to include its installation within the associated docker image.
+A compiled form of `sphgen_cpp` and the binary required for installation of `chimera` are both available within the `dock6_utils` directory of the associated dock6 imge:
```
Comment (davidegraff) [inline README.md:39]:
```suggestion
A compiled form of `sphgen_cpp` and the binary required for installation of `chimera` are both available within the `dock6_utils` directory of the associated dock6 image:
```

File: README.md [hunk 5]
```diff
@@ -23,16 +23,25 @@ This repository contains the source of pyscreener, both a library and software f
...
+* `docker build -t pyscreener:dock6 --target base-dock6 .` : Creates an image from `pyscreener:base` containing utility software needed for `DOCK6` to run once installed 
 
-Note : The above commands assume you are running them within the directory containing both `Dockerfile` and `environment.yml`
+Notes : 
+1. Within the docker container, the environment `base` will e activated by default which contains all the required python dependencies so there is no need to manually activate an environment once inside the container
+2. If installing using docker, then the below installation stages are not required with the exception of `DOCK6` however.
```
Comment (davidegraff) [inline README.md:44]:
```suggestion
2. If installing using docker, then the below installation stages are not required for the corresponding vina-type software. However, the DOCK6 directions must still be followed.
```

**Review by davidegraff — COMMENTED** [review]
just some typos then lgtm

**Comment by Bundaberg-Joey** [comment]
All good points, definitely makes the README clearer / easier to read. Have accepted the changes.

**Review by davidegraff — APPROVED** [review]
